orchagent/vnetorch.cpp

```diff
@@ -1609,39 +1613,115 @@ void VNetRouteOrch::removeBfdSession(const string& vnet, const NextHopKey& endpo
+        info.monitor = monitor_addr;
+        info.state = MONITOR_SESSION_STATE::MONITOR_SESSION_STATE_DOWN;
+    }
+}
+
+void VNetRouteOrch::removeMonitoringSession(const string& vnet, const NextHopKey& endpoint, const IpAddress& monitor_addr, IpPrefix& ipPrefix)
+{
+    SWSS_LOG_ENTER();
+
+    IpAddress endpoint_addr = endpoint.ip_address;
+    if (monitor_info_[vnet].find(ipPrefix) == monitor_info_[vnet].end() ||
+        monitor_info_[vnet][ipPrefix].find(endpoint) == monitor_info_[vnet][ipPrefix].end())
+    {
+        SWSS_LOG_ERROR("Monitor session for prefix %s endpoint %s does not exist", ipPrefix.to_string().c_str(), endpoint_addr.to_string().c_str());
+    }
+
+    string key = ipPrefix.to_string() + ":" + monitor_addr.to_string();
+    monitor_session_producer_->del(key);
+    monitor_info_[vnet][ipPrefix].erase(endpoint);
+}
+
+void VNetRouteOrch::setEndpointMonitor(const string& vnet, const map<NextHopKey, IpAddress>& monitors, NextHopGroupKey& nexthops, const string& monitoring, IpPrefix& ipPrefix)
 {
     SWSS_LOG_ENTER();
 
     for (auto monitor : monitors)
     {
         NextHopKey nh = monitor.first;
         IpAddress monitor_ip = monitor.second;
-        if (nexthop_info_[vnet].find(nh.ip_address) == nexthop_info_[vnet].end())
+        if (monitoring =="custom")
         {
-            createBfdSession(vnet, nh, monitor_ip);
+            if (monitor_info_[vnet].find(ipPrefix) == monitor_info_[vnet].end() ||
+                monitor_info_[vnet][ipPrefix].find(nh) == monitor_info_[vnet][ipPrefix].end())
+            {
+                createMonitoringSession(vnet, nh, monitor_ip, ipPrefix);
+            }
+        }
+        else
+        {
+            if (nexthop_info_[vnet].find(nh.ip_address) == nexthop_info_[vnet].end())
+            {
+                createBfdSession(vnet, nh, monitor_ip);
+            }
+            nexthop_info_[vnet][nh.ip_address].ref_count++;
         }
-
-        nexthop_info_[vnet][nh.ip_address].ref_count++;
     }
 }
 
-void VNetRouteOrch::delEndpointMonitor(const string& vnet, NextHopGroupKey& nexthops)
+void VNetRouteOrch::delEndpointMonitor(const string& vnet, NextHopGroupKey& nexthops, IpPrefix& ipPrefix)
 {
     SWSS_LOG_ENTER();
 
     std::set<NextHopKey> nhks = nexthops.getNextHops();
+    bool is_custom_monitoring = false;
+    if (monitor_info_[vnet].find(ipPrefix) != monitor_info_[vnet].end())
+    {
+        is_custom_monitoring = true;
+    }
     for (auto nhk: nhks)
     {
         IpAddress ip = nhk.ip_address;
-        if (nexthop_info_[vnet].find(ip) != nexthop_info_[vnet].end()) {
-            if (--nexthop_info_[vnet][ip].ref_count == 0)
-            {
-                IpAddress monitor_addr = nexthop_info_[vnet][ip].monitor_addr;
-                removeBfdSession(vnet, nhk, monitor_addr);
+        if (is_custom_monitoring && monitor_info_[vnet][ipPrefix].find(nhk) != monitor_info_[vnet][ipPrefix].end())
```

Comment:
this should be in an outside `if`. What if custom monitoring is true but` find(nhk) == end()`. It will take the BFD path. Please correct.

Reply:
done